File: CZECHMATE/CZECHMATE/Services/PuzzleService.swift
//
//  PuzzleService.swift
//  Lichess veřejné API — denní puzzle (FEN, řešení, rating).
//

import Foundation

/// Stažená denní úloha z Lichess (bez surového JSONu v UI).
struct DailyPuzzle: Sendable, Equatable {
    let id: String
    let fen: String
    let rating: Int?
    /// UCI tahy (např. "e2e4").
    let solutionMoves: [String]
    let themes: [String]
}

enum PuzzleServiceError: LocalizedError {
    case invalidHTTP
    case missingFen
    case decodingFailed

    var errorDescription: String? {
        switch self {
        case .invalidHTTP: return "Neplatná odpověď serveru."
        case .missingFen: return "V odpovědi chybí pozice (FEN)."
        case .decodingFailed: return "Nepodařilo se zpravit data puzzle."
        }
    }
}

/// Načítání puzzlů přes `URLSession` (Lichess public API).
enum PuzzleService {
    private static let dailyURL = URL(string: "https://lichess.org/api/puzzle/daily")!

    private struct LichessDailyResponse: Decodable {
        let puzzle: PuzzlePayload
    }

    private struct PuzzlePayload: Decodable {
        let id: String
        let rating: Int?
        let fen: String?
        let solution: [String]?
        let themes: [String]?
    }

    /// `GET https://lichess.org/api/puzzle/daily`
    static func fetchDailyPuzzle(using session: URLSession = .shared) async throws -> DailyPuzzle {
        var req = URLRequest(url: dailyURL)
        req.timeoutInterval = 25
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse, (200 ... 299).contains(http.statusCode) else {
            throw PuzzleServiceError.invalidHTTP
        }
        let decoded: LichessDailyResponse
        do {
            decoded = try JSONDecoder().decode(LichessDailyResponse.self, from: data)
        } catch {
            throw PuzzleServiceError.decodingFailed
        }
        guard let fen = decoded.puzzle.fen?.trimmingCharacters(in: .whitespacesAndNewlines), !fen.isEmpty else {
            throw PuzzleServiceError.missingFen
        }
        return DailyPuzzle(
            id: decoded.puzzle.id,
            fen: fen,
            rating: decoded.puzzle.rating,
            solutionMoves: decoded.puzzle.solution ?? [],
            themes: decoded.puzzle.themes ?? []
        )
    }
}

File: flutter_czechmate/lib/features/puzzle/puzzle_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

/// Parity with `PuzzleService.swift` — Lichess daily puzzle.
class DailyPuzzle {
  const DailyPuzzle({
    required this.id,
    required this.fen,
    this.rating,
    this.solutionMoves = const [],
    this.themes = const [],
  });

  final String id;
  final String fen;
  final int? rating;
  final List<String> solutionMoves;
  final List<String> themes;
}

class PuzzleService {
  static const _dailyUrl = 'https://lichess.org/api/puzzle/daily';

  static Future<DailyPuzzle> fetchDaily({http.Client? client}) async {
    final c = client ?? http.Client();
    try {
      final req = Uri.parse(_dailyUrl);
      final res = await c.get(req, headers: {'Accept': 'application/json'}).timeout(const Duration(seconds: 25));
      if (res.statusCode < 200 || res.statusCode >= 300) {
        throw Exception('HTTP ${res.statusCode}');
      }
      final map = jsonDecode(utf8.decode(res.bodyBytes));
      if (map is! Map<String, dynamic>) throw Exception('Invalid JSON response');
      final puzzle = map['puzzle'];
      if (puzzle is! Map<String, dynamic>) {
        throw Exception('Daily puzzle payload is missing the puzzle object.');
      }
      final fen = (puzzle['fen'] as String?)?.trim();
      if (fen == null || fen.isEmpty) {
        throw Exception('Daily puzzle is missing a valid FEN position.');
      }
      final sol = (puzzle['solution'] as List<dynamic>?)?.map((e) => '$e').toList() ?? const <String>[];
      final themes = (puzzle['themes'] as List<dynamic>?)?.map((e) => '$e').toList() ?? const <String>[];
      return DailyPuzzle(
        id: puzzle['id'] as String? ?? 'daily',
        fen: fen,
        rating: (puzzle['rating'] as num?)?.toInt(),
        solutionMoves: sol,
        themes: themes,
      );
    } finally {
      if (client == null) {
        c.close();
      }
    }
  }
}
